storm-server/src/main/java/org/apache/storm/nimbus/AssignmentDistributionService.java

```diff
@@ -0,0 +1,312 @@
+ *
+ * <p>We support multiple working threads to distribute assignment, every thread has a queue buffer.
+ *
+ * <p>Master will shuffle its node request to the queues, if the target queue is full, we just discard the request,
+ * let the supervisors sync instead.
+ *
+ * <p>Caution: this class is not thread safe.
+ *
+ * <pre>{@code
+ * Working mode
+ *                      +--------+         +-----------------+
+ *                      | queue1 |   ==>   | Working thread1 |
+ * +--------+ shuffle   +--------+         +-----------------+
+ * | Master |   ==>
+ * +--------+           +--------+         +-----------------+
+ *                      | queue2 |   ==>   | Working thread2 |
+ *                      +--------+         +-----------------+
+ * }
+ * </pre>
+ */
+public class AssignmentDistributionService implements Closeable {
+    private static final Logger LOG = LoggerFactory.getLogger(AssignmentDistributionService.class);
+    private ExecutorService service;
+    /**
+     * Flag to indicate if the service is active.
+     */
+    private volatile boolean active = false;
+
+    private Random random;
+    /**
+     * Working threads num.
+     */
+    private int threadsNum = 0;
+    /**
+     * Working thread queue size.
+     */
+    private int queueSize = 0;
+
+    /**
+     * Assignments request queue.
+     */
+    private volatile Map<Integer, LinkedBlockingQueue<NodeAssignments>> assignmentsQueue;
+
+    /**
+     * local supervisors for local cluster assignments distribution.
+     */
+    private Map<String, Supervisor> localSupervisors;
+
+    private Map conf;
+
+    private boolean isLocalMode = false; // boolean cache for local mode decision
+
+    /**
+     * Function for initialization.
+     *
+     * @param conf config
+     */
+    public void prepare(Map conf) {
+        this.conf = conf;
+        this.random = new Random(47);
+
+        this.threadsNum = ObjectReader.getInt(conf.get(DaemonConfig.NIMBUS_ASSIGNMENTS_SERVICE_THREADS), 10);
+        this.queueSize = ObjectReader.getInt(conf.get(DaemonConfig.NIMBUS_ASSIGNMENTS_SERVICE_THREAD_QUEUE_SIZE), 100);
+
+        this.assignmentsQueue = new HashMap<>();
+        for (int i = 0; i < threadsNum; i++) {
+            this.assignmentsQueue.put(i, new LinkedBlockingQueue<NodeAssignments>(queueSize));
+        }
+        //start the thread pool
+        this.service = Executors.newFixedThreadPool(threadsNum);
+        this.active = true;
+        //start the threads
+        for (int i = 0; i < threadsNum; i++) {
+            this.service.submit(new DistributeTask(this, i));
+        }
+        // for local cluster
+        localSupervisors = new HashMap<>();
+        if (ConfigUtils.isLocalMode(conf)) {
+            isLocalMode = true;
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.active = false;
+        this.service.shutdownNow();
+        try {
+            this.service.awaitTermination(10L, TimeUnit.SECONDS);
```

Comment:
This always will wait the full 10 seconds because there is no way to signal to the DistributeTask that it should shut down.  This is at least in part why it is taking so long for the tests to run.